### PR TITLE
[BACKLOG-21023] - Query a streaming dataservice with a fixed window s…

### PR DESCRIPTION
**Pull request by marcoslarsen**
…ize, update rate and row-based.

@pentaho/kashyyyk @kurtwalker @mkambol  Please review.

### CODE REVIEW

File: impl/src/main/java/org/pentaho/di/trans/dataservice/resolvers/TransientResolver.java [hunk 1]
```diff
@@ -122,6 +124,10 @@ private DataServiceMeta createDataServiceMeta( String dataServiceName ) {
         local = true;
         stepName = stepName.replace( LOCAL, "" );
       }
+      if ( stepName.startsWith( STREAMING ) ) {
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/resolvers/TransientResolver.java]: Is this working after the LOCAL above because of this order of this in [DataServiceMetaFactory](https://github.com/pentaho/pdi-dataservice-server-plugin/pull/279/files#diff-5980933c914e857e00500d96cf032427)?:
`buildTransient( fullFileName, local + stream + step.getName(), rowLimit );`

Shouldn't we add a third case in the DataServiceMetaFactory that adds the LOCAL and STREAMING, and then deal with it in here, something like a third if:
`if ( stepname.startsWith( STREAMING_LOCAL ) ) {`

This way the code doesn't brake, if someone changes the line order for this if here, or the step name in DataServiceMetaFactory.

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  private final String serviceStepName;
+  private final AtomicBoolean isRunning = new AtomicBoolean( false );
+
+  private StreamList<RowMetaAndData> stepStream;
+
+  /**
+   * Service listener cache.
+   */
+  private final Cache<String, StreamExecutionListener> serviceListeners = CacheBuilder.newBuilder()
+    .concurrencyLevel( 4 )
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: this is already the default value, so we can remove it or put a more meaningful value... this number has a performance impact, and should be equal to maximum number of concurrent threads that can access the cache at any given point in time, without incurring in contention. See [javadoc](https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#concurrencyLevel(int)).

Maybe 4 could be a good value (?) but we should put that as a private constant in the class,

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+        removal.getValue().unSubscribe();
+
+        log.logDebug( DataServiceConstants.STREAMING_CACHE_REMOVED + removal.getKey() );
+      }
+    } )
+    .softValues()
+    .build();
+
+  private long lastCacheCleanupMillis = 0;
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: is this lost in here?

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        : stepStream.getStream().buffer( windowMillis, TimeUnit.MILLISECONDS )
+        : stepStream.getStream().buffer( windowSize );
+
+      streamListener = new StreamExecutionListener( buffer );
+
+      serviceListeners.put( cacheId, streamListener );
+    }
+
+    if ( !isRunning.get() ) {
+      isRunning.compareAndSet( false, true );
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: You can just set, since the test was already done in the if statement. 
But if you really want atomic operations you should probably do the compareAndSet and then call startService() depending on the return value of the compareAndSet

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+          }
+        }
+      } );
+    } catch ( KettleException e ) {
+      throw Throwables.propagate( e );
+    }
+  }
+
+  /**
+   * Clean um service listener cache.
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: pt ;)

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java (hunk 3 above)
Comment (marcoslarsen) [reply impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: That's used to control the cleanup calls

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java (hunk 2 above)
Comment (marcoslarsen) [reply impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: To be removed. Stayed there because of testing.

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+          try {
+            if ( log.isRowLevel() ) {
+              log.logRowlevel( DataServiceConstants.PASSING_ALONG_ROW + rowMeta.getString( row ) );
+            }
+          } catch ( KettleValueException e ) {
+            // Ignore error
+          }
+
+          RowMetaAndData rowData = new RowMetaAndData();
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: Shouldn't the rowData field be inside the serviceListeners.size() > 0 if below, since it isn't used outside of that scope?

File: impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    } catch ( KettleException e ) {
+      throw Throwables.propagate( e );
+    }
+  }
+
+  /**
+   * Clean um service listener cache.
+   */
+  private void serviceCacheCleanup() {
+    long currentTime = new Date().getTime();
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java]: Consider using `System.currentTimeMillis` instead, since it has better performance (not much, from `Date().getTime()` I reckon), but it will save on GC execution, by skipping the creation of a new object, that fills memory and then has to be cleaned up.

Same on [StreamingServiceTransExecutor](https://github.com/pentaho/pdi-dataservice-server-plugin/pull/279/files#diff-b6aad85ffbd8f09c0d85e7ff8295f8bf), but more relevant here, since this is called for each row.

File: impl/src/main/java/org/pentaho/di/trans/dataservice/ui/controller/DataServiceDialogController.java [hunk 8]
```diff
@@ -66,6 +66,7 @@ public void init() throws InvocationTargetException, XulException {
 
     XulTextbox serviceName = getElementById( "service-name" );
     XulMenuList<String> steps = getElementById( "trans-steps" );
+
```
Comment (joaormf) [inline impl/src/main/java/org/pentaho/di/trans/dataservice/ui/controller/DataServiceDialogController.java]: No changes really... can't we just revert this change
